### PR TITLE
Use the proxy args in the `FSDPParamUnpaddingVisitor`

### LINKED ISSUE #1092: unhashable type: slice for Thunder and Nous-Hermes-13b
## 🐛 Bug

When running benchmark we get an error:

>  TypeError: unhashable type: \'slice\'

Models affected: 'dolly-v2-3b', 'Nous-Hermes-13b'
### To Reproduce

Please use:

1 node with 8 GPUs.
Image "INTERNAL_IMAGE:pjnl-20240830"

Then execute:
```
torchrun --standalone --max-restarts=0 --no-python --nproc-per-node=8 python /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py \
    --model_name Nous-Hermes-13b \
    --distributed_mode fsdp \
    --shard_mode zero2 \
    --compile thunder \
    --checkpoint_activations False \
    --low_precision_mode none  \
    --micro_batch_size 1
```

### Expected behavior

We should not get this error :) 

### Environment

system.device_product_name                      DGXH100
system.gpu_driver_version                    535.129.03
libraries.cuda                               12.6.1.005
libraries.pip.lightning               2.4.0.dev20240728
libraries.pip.lightning-thunder              0.2.0.dev0
libraries.pip.lightning-utilities                0.11.6
libraries.pip.litgpt                             0.4.11
libraries.pip.nvfuser                 0.2.10+git58dfdc1
libraries.pip.pytorch-lightning                   2.4.0
libraries.pip.torch                  2.5.0a0+git578b8d7
libraries.pip.torchmetrics                        1.4.1
libraries.pip.torchvision              0.19.0a0+d23a6e1

### PR DESCRIPTION
**Pull request by kiya00**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #1092.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kiya00** [comment]
Hi @crcrpar , I debugged on issue 1092, the error is as follows, I tried to fix it by simply using proxy arguments, is that the right way? 
```
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/lightning_thunder-0.2.0.dev0-py3.10.egg/thunder/__init__.py", line 599, in get_computation_and_inputs
[rank1]:     new_prologue_trc, new_computation_trc, new_epilogue_trc = transform.transform_traces_pre_prologue(
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/lightning_thunder-0.2.0.dev0-py3.10.egg/thunder/distributed/transforms/fsdp_v2.py", line 391, in transform_traces_pre_prologue
[rank1]:     new_computation_trace = visitor_transform(
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/lightning_thunder-0.2.0.dev0-py3.10.egg/thunder/core/transforms.py", line 366, in visitor_transform
[rank1]:     visit_type = visit(bsym)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/lightning_thunder-0.2.0.dev0-py3.10.egg/thunder/distributed/transforms/fsdp_v2.py", line 65, in __call__
[rank1]:     if not any(variableify(a) in self.swap_map for a in bsym.flat_args):
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/lightning_thunder-0.2.0.dev0-py3.10.egg/thunder/distributed/transforms/fsdp_v2.py", line 65, in <genexpr>
[rank1]:     if not any(variableify(a) in self.swap_map for a in bsym.flat_args):
[rank1]: TypeError: unhashable type: 'slice'
```

**Comment by crcrpar** [comment]
Sounds right

File: thunder/distributed/transforms/fsdp_v2.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def __call__(self, bsym: BoundSymbol) -> VISIT_TYPE:
             self.swap_map[variableify(padded_tensor)] = unpadded_tensor
             return VISIT_TYPE.INSERT_AFTER
 
-        if not any(variableify(a) in self.swap_map for a in bsym.flat_args):
+        if not any(variableify(a) in self.swap_map for a in bsym.flat_proxy_args):
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:65]: I guess it's reasonable to filter out non-TensorProxy args, even?

**Comment by kiya00** [comment]
Test results:
`torchrun --standalone --max-restarts=0 --no-python --nproc-per-node=8 python /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py \
    --model_name Nous-Hermes-13b \
    --distributed_mode fsdp \
    --shard_mode zero2 \
    --compile thunder \
    --checkpoint_activations False \
    --low_precision_mode none  \
    --micro_batch_size 1
    ` 

```
Model name: Nous-Hermes-13b
Seq Length: 2048
Micro BS: 1
Global BS: 8
Number of Layers: 40
Number of parameters: 1.63B
Distributed Mode: fsdp
Sharding Mode: zero2
Bucketing: none
Compiler: thunder
Low Precision Mode: none
Average iter time: 408.37 ms
Memory used: 53.90 GB
Tokens/s: 40112.05
Tokens/s/GPU: 5014.01
TFLOP/s: 3294.92
```

**Review by crcrpar — APPROVED** [review]
LGTM

**Review by t-vi — APPROVED** [review]
Thank you @kiya00

**Pull request merged**